### PR TITLE
TEST: Added Unit test for test student cannot put teaching period

### PR DESCRIPTION
**Pull request by Guoheng-Liao**
# Description

Modified the variables in setting_test.rb and unified the variable format.

Added unit test for activity type error method.

Added unit test for test student cannot put teaching period.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if appropriate
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have created or extended unit tests to address my new additions
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -26,6 +26,12 @@
   add_index "activity_types", ["abbreviation"], name: "index_activity_types_on_abbreviation", unique: true, using: :btree
   add_index "activity_types", ["name"], name: "index_activity_types_on_name", unique: true, using: :btree
 
+  create_table "ar_internal_metadata", primary_key: "key", force: :cascade do |t|
```
Comment (macite) [inline db/schema.rb]: I think this is left over from the rails upgrade...

Do you really need to add this table to the database for this test change to work?

Comment (macite) [reply db/schema.rb]: I suggest you revert the changes to this file.

In the future take care with what you are adding to each commit.

File: test/models/learning_outcome_test.rb [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      }
+    @unit = Unit.create(data)
+  end
+
+  # Check abbreviation uniqueness
+  def test_create_learning_outcome_with_unique_abbrev
+    #Create new unit.
+    learning_outcome_count = LearningOutcome.count
+    unit_1 = {
+        unit_id: 'ASD10001',
```
Comment (macite) [inline test/models/learning_outcome_test.rb]: Unit id is a relation to a unit - not the unit code. This wont work as you expect.

Try adding a simple test that you can actually create a learning outcome and that it is then available within the unit.

File: test/models/learning_outcome_test.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    learning_outcome2 = LearningOutcome.create!(unit_2)
+    
+    #Check the abbreviation is not empty.
+    refute_empty learning_outcome.abbreviation    
+    refute_empty learning_outcome2.abbreviation
+
+    #Check the abbreviation of the unit is unique.
+    refute_equal learning_outcome, learning_outcome2
```
Comment (macite) [inline test/models/learning_outcome_test.rb]: This isn't checking abbreviations...

File: test/models/learning_outcome_test.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        unit_id: 'ASD10002',
+        ilo_number: learning_outcome_count + 1,
+        abbreviation: 'PR',
+        name: 'Programming',
+        description: 'Create programming'
+    }
+
+    learning_outcome2 = LearningOutcome.create!(unit_2)
+    
+    #Check the abbreviation is not empty.
```
Comment (macite) [inline test/models/learning_outcome_test.rb]: Why would it be empty?
Is it valid?
Check details about validations in rails.

File: test/models/learning_outcome_test.rb (hunk 3 above)
Comment (macite) [reply test/models/learning_outcome_test.rb]: I'll stop here...

I suggest you make a very simple test first to demonstrate that the basics of learning outcomes are tested. Create the learning outcome and then test that everything about it is setup correctly, and that you can interact with it via its relationships etc.

**Comment by macite** [comment]
Also - please work on your PR description. What you have provided does not tell me what this PR does.

File: db/schema.rb (hunk 1 above)
Comment (macite) [reply db/schema.rb]: These have not been addressed.

File: test/api/settings_test.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    include TestHelpers::AuthHelper
+    include TestHelpers::JsonHelper
+
+    def app
+        Rails.application
+    end
+
+    # Get config details
+    def test_get_config_details        
+        expected_ProductName =  Doubtfire::Application.config.institution[:product_name]
```
Comment (macite) [inline test/api/settings_test.rb]: Variable name case! This is very sloppy.

File: test/api/settings_test.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        # Check returned details match as expected
+        assert_equal expected_privacy, returned_privacy
+        assert_equal expected_plagiarism, returned_plagiarism
+    end
+
+    # Get all unit teaching periods info
+    def test_get_all_unit_teaching_periods_info
+        # The GET we are testing
+        get '/api/teaching_periods/id/units'
```
Comment (macite) [inline test/api/settings_test.rb]: Do we need this API endpoint? It does not sound useful to me. Can you check where it is used in the front end and recommend we delete it if it is not used.

File: test/api/settings_test.rb (hunk 5 above)
Comment (Guoheng-Liao) [reply test/api/settings_test.rb]: Hi Andrew, I have modified variables name case, and unified their format.

File: test/api/settings_test.rb (hunk 6 above)
Comment (Guoheng-Liao) [reply test/api/settings_test.rb]: I have checked the api endpoint, and as you said we don't need it. 
So do I need to delete this link (get '/ api / teaching_periods / id / units') and its test method (def test_get_all_unit_teaching_periods_info) and commit setting_test.rb again?

File: test/api/activity_types_api_test.rb [hunk 7]
```diff
@@ -50,4 +50,50 @@ def test_put_activity_types
     first_activity_type = ActivityType.first
     assert_json_matches_model(last_response_body, first_activity_type, response_keys)
   end
+
+  def test_post_activity_types_cannot_auth
```
Comment (linhhtr) [inline test/api/activity_types_api_test.rb]: Please change the test name into `test_student_cannot_post_activity_type` for clearer description

File: test/api/activity_types_api_test.rb [hunk 8]
```diff
@@ -50,4 +50,50 @@ def test_put_activity_types
...
+    post_json '/api/activity_types', with_auth_token(data_to_post, user)
+
+    # Check if the post does not get through
+    assert_equal 403, last_response.status  
+
+    # Check if the number of activity type is the same as initially
+    assert_equal ActivityType.count, number_of_activity_type
+  end  
+
+  def test_put_activity_types_cannot_auth
```
Comment (linhhtr) [inline test/api/activity_types_api_test.rb]: Also change the name of this test

File: test/api/activity_types_api_test.rb [hunk 9]
```diff
@@ -50,4 +50,50 @@ def test_put_activity_types
...
+    user = FactoryBot.build(:user, :student)
+
+    # Create a dummy activity type
+    data_to_put = {
+      activity_type: FactoryBot.build(:activity_type),
+      auth_token: auth_token
+    }
+
+    # Update activity_type with id = 1, but the student user does not have permissions to put it.
+    put_json '/api/activity_types/1', with_auth_token(data_to_put, user)
```
Comment (linhhtr) [inline test/api/activity_types_api_test.rb]: Please do not put `data_to_put` into the activity type of which id is explicitly `1`. Let create a new activity type by FactoryBot `activity_type = FactoryBot.create(:activity_type)` then assign `data_to_put`into to replace that newly created `activity_type` of which id can be called by `activity_type.id`. If this does not make sense to you, please ask

**Comment by Guoheng-Liao** [comment]
@macite Hello Andrew, I have updated the variable names in setting_test_api.rb and unified their format.
Also, I added two new unit tests, They are used to test that students cannot post or put new activity types.

File: test/api/settings_test.rb (hunk 6 above)
Comment (macite) [reply test/api/settings_test.rb]: Sorry, I thought I had replied to this earlier. Yes, please remove the test and the end point.

Comment (macite) [reply test/api/settings_test.rb]: Once that is done I can merge the PR.

**Comment by Guoheng-Liao** [comment]
@macite Hello Andrew, I removed the test and code we discussed a while ago.

**Pull request merged**